fdbrpc/RESTUtils.actor.cpp

```diff
@@ -0,0 +1,276 @@
+}
+
+RESTUrl::RESTUrl(const std::string& fullUrl, const std::string& b, const bool isSecure) : body(b) {
+	parseUrl(fullUrl, isSecure);
+}
+
+void RESTUrl::parseUrl(const std::string& fullUrl, const bool isSecure) {
+	// Sample valid URIs
+	// 1. With 'host' & 'resource' := '<protocol>://<host>/<resource>'
+	// 2. With 'host', 'service' & 'resource' := '<protocol>://<host>:port/<resource>'
+	// 3. With 'host', 'service', 'resource' & 'reqParameters' := '<protocol>://<host>:port/<resource>?<parameter-list>'
+
+	try {
+		StringRef t(fullUrl);
+		StringRef p = t.eat("://");
+		std::string protocol = p.toString();
+		boost::algorithm::to_lower(protocol);
+		if (!isProtocolSupported(protocol)) {
+			throw format("Invalid REST URI protocol '%s'", protocol.c_str());
+		}
+
+		// Ensure connection secure knob setting matches with the input URI
+		if ((isSecurePrototol(protocol) && !isSecure) || (!isSecurePrototol(protocol) && isSecure)) {
+			throw format("Invalid REST URI protocol secure knob '%s'", fullUrl.c_str());
+		}
+
+		// extract 'resource' and optional 'parameter list' if supplied in the URL
+		uint8_t foundSeparator = 0;
+		StringRef hostPort = t.eatAny("/?", &foundSeparator);
+		if (foundSeparator == '/') {
+			resource = t.eat("?").toString();
+			reqParameters = t.eat().toString();
+		}
+
+		// hostPort is at least a host or IP address, optionally followed by :portNumber or :serviceName
+		StringRef hRef(hostPort);
+		StringRef h = hRef.eat(":");
+		if (h.size() == 0) {
+			throw std::string("host cannot be empty");
+		}
+		host = h.toString();
+		service = hRef.eat().toString();
+
+		TraceEvent("RESTClient_ParseURI")
+		    .detail("URI", fullUrl)
+		    .detail("Host", host)
+		    .detail("Service", service)
+		    .detail("Resource", resource)
+		    .detail("ReqParameters", reqParameters);
+	} catch (std::string& err) {
+		TraceEvent("RESTClient_ParseError").detail("URI", fullUrl).detail("Error", err);
+		throw rest_invalid_uri();
+	}
+}
+
+// Only used to link unit tests
+void forceLinkRESTUtilsTests() {}
+
+TEST_CASE("fdbrpc/RESTUtils") {
+	// invalid protocol
+	try {
+		std::string uri("httpx://foo/bar");
+		RESTUrl r(uri, false);
+		ASSERT(false);
+	} catch (Error& e) {
+		if (e.code() != error_code_rest_invalid_uri) {
+			throw e;
+		}
+	}
+
+	// mismatch protocol and knob values
+	try {
+		std::string uri("http://foo/bar");
+		RESTUrl r(uri, true);
+		ASSERT(false);
+	} catch (Error& e) {
+		if (e.code() != error_code_rest_invalid_uri) {
+			throw e;
+		}
+	}
+
+	// missing host
+	try {
+		std::string uri("https://:/bar");
+		RESTUrl r(uri, true);
+		ASSERT(false);
+	} catch (Error& e) {
+		if (e.code() != error_code_rest_invalid_uri) {
+			throw e;
+		}
+	}
+
+	// valid URI with service
+	try {
+		std::string uri("https://host:80/foo/bar");
+		RESTUrl r(uri, true);
+		ASSERT_EQ(r.host.compare("host"), 0);
+		ASSERT_EQ(r.service.compare("80"), 0);
+		ASSERT_EQ(r.resource.compare("foo/bar"), 0);
+	} catch (Error& e) {
+		throw e;
```

Comment:
Why try/catch just to throw?

Reply:
Seems like fallout from debugging; shall remove it. Thanks